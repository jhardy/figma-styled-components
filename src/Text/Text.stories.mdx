import { SectionTitle } from "../SectionTitle";
import { Text } from "./Text";
import { Label } from "../Label";
import { Story, Meta } from "@storybook/addon-docs/blocks";

<Meta title="MDX|Typography" component={Text} />

# Typography Components

## Text

<Story name="Text">
  <div class="component-box">
    <div className="text-components">
      <div class="text-wrapper">
        <Text size="xlarge">Positive Application</Text>
        <div className="text-block">
          <ul>
            <li>
              <Text>UI11 Normal</Text>
            </li>
            <li>
              <Text weight="medium">UI11 Medium</Text>
            </li>
            <li>
              <Text weight="bold">UI11 Bold</Text>
            </li>
          </ul>
          <ul>
            <li>
              <Text size="medium">UI12 Normal</Text>
            </li>
            <li>
              <Text size="medium" weight="medium">
                UI12 Medium
              </Text>
            </li>
            <li>
              <Text size="medium" weight="bold">
                UI12 Bold
              </Text>
            </li>
          </ul>
          <ul>
            <li>
              <Text size="large">UI13 Normal</Text>
            </li>
            <li>
              <Text size="large" weight="medium">
                UI13 Medium
              </Text>
            </li>
            <li>
              <Text size="large" weight="bold">
                UI13 Bold
              </Text>
            </li>
          </ul>
          <ul>
            <li>
              <Text size="xlarge">UI14 Normal</Text>
            </li>
            <li>
              <Text size="xlarge" weight="medium">
                UI14 Medium
              </Text>
            </li>
            <li>
              <Text size="xlarge" weight="bold">
                UI14 Bold
              </Text>
            </li>
          </ul>
        </div>
      </div>
      <div class="text-wrapper">
        <Text size="xlarge">Negative Application</Text>
        <div className="text-block negative-text">
          <ul>
            <li>
              <Text inverted={true}>UI11 Normal</Text>
            </li>
            <li>
              <Text inverted={true} weight="medium">
                UI11 Medium
              </Text>
            </li>
            <li>
              <Text inverted={true} weight="bold">
                UI11 Bold
              </Text>
            </li>
          </ul>
          <ul>
            <li>
              <Text inverted={true} size="medium">
                UI12 Normal
              </Text>
            </li>
            <li>
              <Text inverted={true} size="medium" weight="medium">
                UI12 Medium
              </Text>
            </li>
            <li>
              <Text inverted={true} size="medium" weight="bold">
                UI12 Bold
              </Text>
            </li>
          </ul>
          <ul>
            <li>
              <Text inverted={true} size="large">
                UI13 Normal
              </Text>
            </li>
            <li>
              <Text inverted={true} size="large" weight="medium">
                UI13 Medium
              </Text>
            </li>
            <li>
              <Text inverted={true} size="large" weight="bold">
                UI13 Bold
              </Text>
            </li>
          </ul>
          <ul>
            <li>
              <Text inverted={true} size="xlarge">
                UI14 Normal
              </Text>
            </li>
            <li>
              <Text inverted={true} size="xlarge" weight="medium">
                UI14 Medium
              </Text>
            </li>
            <li>
              <Text inverted={true} size="xlarge" weight="bold">
                UI14 Bold
              </Text>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</Story>

| Prop Name | Type                                 | Default  | Required | Description                    |
| :-------- | :----------------------------------- | :------: | :------: | :----------------------------- |
| size      | "small", "medium", "large", "xlarge" | "small"  |    -     | Text's font size               |
| weigth    | "normal", "medium", "bold"           | "normal" |    -     | Text's weight                  |
| inverted  | boolean                              |  false   |    -     | If text is on inverted surface |

## Section Title

<Story name="SectionTitle">
  <div class="component-box">
    <SectionTitle>I am a Section Title</SectionTitle>
  </div>
</Story>

## Label

<Story name="Label">
  <div class="component-box">
    <Label>I am a label</Label>
  </div>
</Story>
