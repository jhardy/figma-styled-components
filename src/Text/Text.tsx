import styled, { css } from 'styled-components'

export type TextSizes = 'small' | 'medium' | 'large' | 'xlarge'
export type TextWeights = 'normal' | 'medium' | 'bold'

export interface TextProps {
  size?: TextSizes
  weight?: TextWeights
  inverted?: boolean
}
   // tslint:disable:object-literal-sort-keys
const textAttributes = {
  small: {
    fontSize: `11px`,
    lineHeight: `16px`,

    letterSpacing: { positive: `0.005em`, negative: `.01em` }
  },
  medium: {
    fontSize: `12px`,
    lineHeight: `16px`,
    letterSpacing: { positive: `0`, negative: `.005em;` }
  },
  large: {
    fontSize: `13px`,
    lineHeight: `24px`,
    letterSpacing: { positive: `-.0025em`, negative: `.0025em` }
  },
  xlarge: {
    fontSize: `14px`,
    lineHeight: `24px`,
    letterSpacing: { positive: `-.001em`, negative: `-.001em` }
  }
}

const getTextWeight = (weight: TextWeights) => {
  switch (weight) {
    case 'bold':
      return 600
    case 'medium':
      return 500
    default:
      return 400
  }
}

const getTextStyles = (
  size: TextSizes = 'small',
  weight: TextWeights = 'normal',
  onNegative: boolean = false
) => {
  const textStyles = textAttributes[size]
  const variant = onNegative ? 'negative' : 'positive'

  return css`
    font-weight: ${getTextWeight(weight)};
    font-size: ${textStyles.fontSize};
    letter-spacing: ${textStyles.letterSpacing[variant]};
    line-height: ${textStyles.lineHeight};
  `
}

export const Text = styled.span<TextProps>`
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter", sans-serif;
  color: ${(props) => (props.inverted ? '#ffffff' : 'rgba(0, 0, 0, 0.8)')};
  ${(props) => getTextStyles(props.size, props.weight, props.inverted)}
`
