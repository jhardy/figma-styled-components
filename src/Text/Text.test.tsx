import 'jest-styled-components'
import * as React from 'react'
import * as renderer from 'react-test-renderer'

import { Text } from './Text'


test('A default Text', () => {
  const tree = renderer.create(<Text>Hello</Text>).toJSON()
  expect(tree).toMatchSnapshot()
})


test('Text can be resized', () => {
  const tree = renderer.create(<Text size='ui12'>Hello</Text>).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Text can have a different weight', () => {
  const tree = renderer.create(<Text weight='bold'>Hello</Text>).toJSON()
  expect(tree).toMatchSnapshot()
})


test('Text can be on negative', () => {
  const tree = renderer.create(<Text size='ui13' inverted={true}>Hello</Text>).toJSON()
  expect(tree).toMatchSnapshot()
})



