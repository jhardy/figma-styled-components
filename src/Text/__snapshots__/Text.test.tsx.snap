// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A default Text 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter",sans-serif;
  color: rgba(0,0,0,0.8);
  font-weight: 400;
  font-size: 11px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  line-height: 16px;
}

<span
  className="c0"
>
  Hello
</span>
`;

exports[`Text can be on negative 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter",sans-serif;
  color: #ffffff;
  font-weight: 400;
  font-size: 13px;
  -webkit-letter-spacing: .0025em;
  -moz-letter-spacing: .0025em;
  -ms-letter-spacing: .0025em;
  letter-spacing: .0025em;
  line-height: 24px;
}

<span
  className="c0"
  size="large"
>
  Hello
</span>
`;

exports[`Text can be resized 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter",sans-serif;
  color: rgba(0,0,0,0.8);
  font-weight: 400;
  font-size: 12px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 16px;
}

<span
  className="c0"
  size="medium"
>
  Hello
</span>
`;

exports[`Text can have a different weight 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter",sans-serif;
  color: rgba(0,0,0,0.8);
  font-weight: 600;
  font-size: 11px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  line-height: 16px;
}

<span
  className="c0"
>
  Hello
</span>
`;
